### PR TITLE
Introduce used in filter

### LINKED ISSUE #1462: Introduce Used In filter
Story #1291 

Add `Used In` filter to the media gallery to allow the user to filter by the area where the images are used in:
- Pages
- Categories
- Blocks
- Products

The filter input type should be the same as the print screen below and allow the user to select multiple options.

Screenshot 2020-06-16 at 13 56 06

**Technical details**
https://github.com/magento/adobe-stock-integration/wiki/MediaGalleryUi-DataProvider

### PR DESCRIPTION
**Pull request by ghost**
### Description (*)

Add Used In filter to the media gallery to allow the user to filter by the area where the images are used in: pages, categories, blocks, products.

### Fixed Issues (if relevant)

1. Fixes magento/adobe-stock-integration#1462: Introduce Used In filter

### Manual testing scenarios (*)

in order to test the search must put images on these area pages, categories, blocks, products.
1. Admin > content > media gallery - click filter button and select your search.

### CODE REVIEW

**Comment by ghost** [comment]
@magento run all tests

File: MediaGalleryUi/Model/GetDetailsByAssetId.php [hunk 1]
```diff
@@ -85,6 +85,7 @@ public function execute(int $assetId): array
             'image_url' => $this->getUrl($asset->getPath()),
             'title' => $asset->getTitle(),
             'path' => $asset->getPath(),
+            'description' => $asset->getDescription(),
```
Comment (sivaschenko) [inline MediaGalleryUi/Model/GetDetailsByAssetId.php]: Looks like this PR contains the changes from https://github.com/magento/adobe-stock-integration/pull/1512. Can you please ensure each PR contains only the changes related to the task

**Review by sivaschenko — CHANGES_REQUESTED** [review]

File: AdobeStockImageAdminUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\AdobeStockImageAdminUi\Model\SearchCriteria\CollectionProcessor\FilterProcessor;
```
Comment (sivaschenko) [inline AdobeStockImageAdminUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php]: Please move the files to `MediaGalleryUi` module.

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Thanks for the PR @gregbalonzo ! The implementation looks good, however, it should be moved to the MediaGalleryUi module

File: AdobeStockImageAdminUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php (hunk 2 above)
Comment (ghost) [reply AdobeStockImageAdminUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php]: I tried moving the file AdobeStockImageAdminUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php
to
stock-integration/MediaGalleryUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/
i cannot get to work

Comment (sivaschenko) [reply AdobeStockImageAdminUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php]: I meant that all the implementation should be moved to MediaGalleryUi module, not just a single file (for sure, the namespaces should be changed accordingly).

Comment (ghost) [reply AdobeStockImageAdminUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php]: Hi sergii, done moving the implementation MediaGalleryUi module

**Comment by gabrieldagama** [comment]
@magento run all tests

File: MediaGalleryUi/Ui/Component/Listing/Columns/Layout/Options.php [hunk 3]
```diff
@@ -0,0 +1,42 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+declare(strict_types=1);
+
+namespace Magento\MediaGalleryUi\Ui\Component\Listing\Columns\Layout;
```
Comment (gabrieldagama) [inline MediaGalleryUi/Ui/Component/Listing/Columns/Layout/Options.php]: Let's change the namespace/folders.

```suggestion
namespace Magento\MediaGalleryUi\Ui\Component\Listing\Filters\Options;
```

File: MediaGalleryUi/Ui/Component/Listing/Columns/Layout/Options.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+use Magento\Framework\Data\OptionSourceInterface;
+use Magento\Framework\Option\ArrayInterface;
+
+/**
+ * Image source filter options
+ */
+class Options implements ArrayInterface
```
Comment (gabrieldagama) [inline MediaGalleryUi/Ui/Component/Listing/Columns/Layout/Options.php]:
```suggestion
class EntityType implements ArrayInterface
```

File: MediaGalleryUi/Ui/Component/Listing/Columns/Layout/Options.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+{
+    /**
+     * @inheritdoc
+     */
+    public function toOptionArray(): array
+    {
+        return [
```
Comment (gabrieldagama) [inline MediaGalleryUi/Ui/Component/Listing/Columns/Layout/Options.php]: Can we inject this array here via di.xml? That will allow other extensions to easily add their entity to this.

**Review by gabrieldagama — CHANGES_REQUESTED** [review]

File: MediaGalleryUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php [hunk 6]
```diff
@@ -0,0 +1,23 @@
+<?php
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
```
Comment (coderimus) [inline MediaGalleryUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php:7]: Please, add `declare(strict_types=1);`

File: MediaGalleryUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+namespace Magento\MediaGalleryUi\Model\SearchCriteria\CollectionProcessor\FilterProcessor;
+
+use Magento\Framework\Api\Filter;
+use Magento\Framework\Api\SearchCriteria\CollectionProcessor\FilterProcessor\CustomFilterInterface;
+use Magento\Framework\Data\Collection\AbstractDb;
+
```
Comment (coderimus) [inline MediaGalleryUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php:15]: The class should have a description of its role.

File: MediaGalleryUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+class EntityType implements CustomFilterInterface
+{
+    /**
+     * @inheritDoc
+     */
+    public function apply(Filter $filter, AbstractDb $collection)
```
Comment (coderimus) [inline MediaGalleryUi/Model/SearchCriteria/CollectionProcessor/FilterProcessor/EntityType.php]:
```suggestion
    public function apply(Filter $filter, AbstractDb $collection): bool
```

**Review by coderimus — CHANGES_REQUESTED** [review]
Hi @gregbalonzo , thank you for your contribution. Please, check my suggestions below.

**Comment by ghost** [comment]
@magento run all tests

**Comment by coderimus** [comment]
Hello @gregbalonzo ! Thank you for the provided changes. Please, fix the conflict at the `image-details.html`

**Comment by ghost** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Review by gabrieldagama — DISMISSED** [review]

**Review by coderimus — DISMISSED** [review]

**Review by chalov-anton — CHANGES_REQUESTED** [review]
:x: QA Failed

Please observe the following points from less important to more important:

- Since plural is used with Block**s**, Categorie**s**, Product**s**, please use plural for Page**s** also
used_in_filter_pages

- there is a frontend issue when applying filters NOT from Standalone Media Gallery, for example from _Catalog - Categories_ 
used_in_filter

- if there is an image which is used in two entities, for example in Product and in Category, and both filters are applied - an _Item (Magento\Framework\View\Element\UiComponent\DataProvider\Document) with the same ID "1" already exists._ text is displayed
used_in_filter_cat_cat
works properly when only one of the filters is applied

**Comment by chalov-anton** [comment]
Test on the Cucumber Studio - https://studio.cucumber.io/projects/131313/test-plan/folders/1320712/scenarios/4930844

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Review by gabrieldagama — DISMISSED** [review]

**Comment by chalov-anton** [comment]
✔️ QA Passed

**Pull request merged**